File: no_style.js
/******************************************************************************
File Name:    no_style.js
Description:  This term frequency program is to read a text file. From there,
              it is to count the frequency of each term in the file while
              taking into consideration the predeterminded list of stop words
              established.
******************************************************************************/


//Reviewes the input from the command line and checks for the correct number of arguments
if (process.argv.length < 3) {
    console.log('Unable to execute term frequency program successfully' + '\n' +
    'Please enter the appropriate arguments in order to run');
    process.exit(1);
}


//Reads text file specified from command line and checks if an error is thrown from reading the file
var fs = require('fs');
var file = __dirname + '/' + process.argv[2];
fs.readFile(file, 'utf8', function (err, data) { 
    if (err) {
        throw err;
    }
    
    
    //Sorts words from text file into an array of Objects where the highest frequency words appear first
    var wordsArray = splitWords(data);
    var wordsObject = createWordsObject(wordsArray);
    var sortedWordsArray = sortWordsObject(wordsObject);
    
    var i = 0;
    
    
    //Prints all words from text file by highest frequency if number of words is less than 25
    if (sortedWordsArray.length < 25) {
        for (i; i < sortedWordsArray.length; i++) {
            console.log(sortedWordsArray[i].word + "  -  " + sortedWordsArray[i].frequency);
        }
    }
    
    
    //Prints the 25 highest frequency words from text file
    else {
        for (i; i < 25; i++) {
            console.log(sortedWordsArray[i].word + "  -  " + sortedWordsArray[i].frequency);
            }
        }
    });


/******************************************************************************
Function Name:  splitWords
Purpose:        This function splits a text file by words into an array
Description:    This function reads the text file specified by the user from
                the command line. It then converts all characters to lower case
                and replaces all non-alphanumeric characters with a space. The
                text is then split by whitespace characters.
Input:          textfile  - the text file specified by the user
Result:         An array of all words contained within the text file
******************************************************************************/
function splitWords (textFile) {
    var wordsArray = [];
    wordsArray = textFile.toLowerCase().replace(/[^a-zA-Z0-9]/g, ' ').split(/\s+/);
    return wordsArray;
}


/******************************************************************************
Function Name:  createWordsObject
Purpose:        This function creates an Object to store the valid words and
                their frequency in an unformatted manner from an array
Description:    This function first reads the list of stop words specified for
                the exercise and stores them into an array. From there, an
                Object is created to store each unique word and its frequency
                from a given array, ensuring that the word is not a stop word.
Input:          wordsArray  - the array of all words found within the text file
Result:         An Object containing the list of valid words and their
                frequencies from an array
******************************************************************************/
function createWordsObject (wordsArray) {
    
    
    //Reads the list of stop words associated to exercise and splits them into an array
    var fs = require('fs');
    var stopWordsFile = __dirname + '/stop_words.txt';
    var readStopWordsFile = [];
    readStopWordsFile = fs.readFileSync(stopWordsFile, 'utf8');
    var splitStopWords = [];
    splitStopWords = readStopWordsFile.split(',');
    
    
    var wordsObject = {};
    
    
    /*
     * Loops through each elemant in the words array and checks to see if the word is not a stop word
     * In addition, it checks to see if the word is not the string "s"
     * The second condition is used to cover the manner in which words originally containing the text "'s" were split
     * Examples of these words include "Bennet's", "Elizabeth's", and "Jame's"
     */
    wordsArray.forEach(function (arrayElement) {
        if (!splitStopWords.includes(arrayElement) && arrayElement !== 's') {
            
            
            /*
             * Checks to see if the Object has a property containing a specific word
             * If it does, the frequency of that property is increated
             * If not, the frequency of that property is set to 1
             */
            if (wordsObject.hasOwnProperty(arrayElement)) {
                wordsObject[arrayElement]++;
            }
            else {
                wordsObject[arrayElement] = 1;
            }
        }
    });
    
    
    return wordsObject;
}


/******************************************************************************
Function Name:  sortWordsObject
Purpose:        This function creates an array of Objects and sorts the Objects
                by the words with the highest frequencies
Description:    This function first creates an array to store the Objects. It
                then formats the Objects in a way to better understand what
                information is being stored into the array. Finally, the array
                is sorted by highest frequency.
Input:          wordsObject  - the Object containing the terms and their
                              frequencies in an unformatted manner
Result:         An array of Objects sorted by the highest frequency words
******************************************************************************/
function sortWordsObject (wordsObject) {
    var sortedWordsArray = [];
    
    
    /*
     * Creates an array of the names of each property within the Object
     * In this case, the names of each property is our valid word from the text file
     * From there, the array is reformatted where the results sort the word and it's frequeny into an understandable interpretation
     */
    sortedWordsArray = Object.keys(wordsObject).map(function (element) {
        return {
            word: element,
            frequency: wordsObject[element]
        };
    });
    
    
    //Sorts the array by the highest frequency words
    sortedWordsArray.sort( function (a, b) {
        return b.frequency - a.frequency;
    });
    
    
    return sortedWordsArray;
}

File: no_style.py
"""
File Name:   no_style.py
    
Description: This program takes one text file as an argument and ouputs the frequencies of valid tokens it contains.
             Valid tokens are considered as a sequence of alphanumeric characters along with a few special characters [a-zA-Z0-9@#*&'].
             The tokens are then reduced to their lowercase form where their frequencies can be computed and shown to the user.
             These alphanumeric sequencies must have a length of at least 2 to be considered valid (i.e., len(token) >= 2).
"""
#!/usr/bin/env python3

import re
import string
import sys

from collections import Counter
from os import path

class Tokenizer:
    """
    The Tokenizer object allows us to compute the frequencies of the tokens within the text file
    
    Attributes:
        tokens (list): Stores all tokens present within the file
        frequencies (Counter): Dictionary subclass that stores the frequency of tokens
        file (str): Stores the name of the file and its location
    """
    def __init__(self, textFile):
        self.tokens = []
        self.frequencies = Counter()
        self.file = textFile
    
    def readChunk(self, fileObject):
        """
        The readChunk method reads a file object in sections of 4096 bytes and returns a generator of the text
        contained wthin this section size
        
        Args:
            fileObject (file object): A file object of the text file being read
        
        Returns:
            A generator for the text equal to the size of the bytes indicated to read
        """
        while True:
            chunk = fileObject.read(4096)
            if not chunk:
                break
            yield chunk
        
    def tokenize(self):
        """
        The tokenize method reads the text file provided and adds each valid token into the 'tokens' attribute
        
        Note:
            The runtime for this method is O(n) where n is the number characters in the text file.
            This is because the method grabs each line in the file and converts all uppercase characters to lowercase.
            From there, it uses a regex expression to find all valid tokens in the file.
            This requires us to parse through each character in the file to grab the list of valid tokens.
            
        Raises:
            OSError: If file cannot be opened
        """
        # Checks that the text file provided can be opened
        try:
            file = open(self.file, encoding='utf-8', errors='replace')
        
        # Displays the error that arose if the file could not be opened
        except OSError as error:
            print(error)
            
        # Performs our tokenization if the file can be opened
        else:
            
            # Creates a regular expression in order to tokenize the file
            pattern = re.compile('\w+')
            
            # Creates a set of characters used for token pattern determination
            charList = list(string.ascii_letters) + list(string.digits) + ['_']
            charSet = set(charList)
            
            # Variables to be used for retaining information on characters and tokens found
            firstChar = ''
            lastChar = ''
            lastToken = ''
            
            # Iterates through each chunk in the file
            for chunk in self.readChunk(file):
                
                # Saves the first character in the chunk and finds all valid tokens through the regular expression created
                firstChar = chunk[0]
                validTokens = pattern.findall(chunk.lower())
                
                # Checks if the first character in current chunk should be part of the last token determined in previous chunk
                # If true, combines last token with first token in current chunk
                # Otherwise, adds last token to list of valid tokens found
                if lastChar in charSet and firstChar in charSet:
                    validTokens[0] = lastToken + validTokens[0]
                elif lastToken:
                    self.tokens.append(lastToken)
                
                # Pops the last valid token found and adds the list of valid tokens to the 'tokens' attribute
                # Saves the last character in the chunk to be used for reference for next chunk to be read
                lastToken = validTokens.pop()
                self.tokens.extend(validTokens)
                lastChar = chunk[-1]
            
            # Adds last token to list of valid characters found
            self.tokens.append(lastToken)
        
        # Closes the file regardless of whether or not an error was thrown
        finally:
            file.close()
        
    def computeWordFrequencies(self):
        """
        The computeWordFrequencies method checks the list of tokens and computes the frequences of each one into the 'frequencies'
        attribute of the object
        
        Note:
            The runtime for this method is O(m) where n is the number of items in the 'tokens' attribute.
            This is because the method grabs each valid token that was in the file.
            From there, the token is added to the dictionary subclass and assigned a value of 1.
            If the token is already present in the datatype, its value is increased by 1.
            
            Counter datatype information: https://docs.python.org/3.6/library/collections.html#collections.Counter
        """
        self.frequencies = Counter(self.tokens)
        
    def print(self):
        """
        The print method displays the list of tokens and their frequencies to the user
        
        Note:
            The runtime for this method is O(p log p) where p is the number of key:value pairs in the 'frequencies' attribute.
            This is because our method grabs each pairing in the dictionary subclass.
            From there, it prints the key as well as its associated value for the user to view through the terminal.
            The pairs are presented in sorted order based on the frequency of the token.
        """
        with open('stop_words.txt') as file:
            stop_words = set(re.split(',', file.read()))
            stop_words.add('s')
        
        count = 0
        for (token, frequency) in self.frequencies.most_common():
            if count >= 25:
                break
            if not token in stop_words:
                print(token + ' - ' + str(frequency))
                count += 1
        
def main():
    """
    The main function runs the tokenizer program and displays the frequencies of each token within the text file
    
    Raises:
        SystemExit: If improper arguments are provided to run the program
    """
    # Checks that the user provides the valid number of arguments and a valid text file to run the program
    if (len(sys.argv) != 2):
        sys.exit('Invalid arguments provided\nPlease provide valid arguments in order to run tokenizer')
    elif (not path.isfile(sys.argv[1]) or sys.argv[1][-4:] != '.txt'):
        sys.exit('Argument provided is not a valid text file\nPlease provide a valid text file in order to run tokenizer')
        
    # Creates a Tokenizer object based on text file given and computes the frequencies of tokens within then file
    tkzr = Tokenizer(sys.argv[1])
    tkzr.tokenize()
    tkzr.computeWordFrequencies()
    tkzr.print()

if __name__ == "__main__":
    main()
